expression/builtin_control.go

```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *caseWhenFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *caseWhenFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinCaseWhenSig{newBaseBuiltinFunc(args, ctx)}
+	l := len(args)
+	// Fill in each 'THEN' clause parameter type.
+	fieldTps := make([]*types.FieldType, 0, (l+1)/2)
+	decimal, flen, isBinaryStr := args[1].GetType().Decimal, 0, false
+	for i := 1; i < l; i += 2 {
+		fieldTps = append(fieldTps, args[i].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[i].GetType().Decimal), mathutil.Max(flen, args[i].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[i].GetType())
+	}
+	if l%2 == 1 {
+		fieldTps = append(fieldTps, args[l-1].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[l-1].GetType().Decimal), mathutil.Max(flen, args[l-1].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[l-1].GetType())
+	}
+	fieldTp := types.AggFieldType(fieldTps)
+	classType := types.AggTypeClass(fieldTps, &fieldTp.Flag)
+	// Set retType to BINARY(0) if all arguments are of type NULL
+	if fieldTp.Tp == mysql.TypeNull {
+		fieldTp.Flen, fieldTp.Decimal = 0, 0
+		types.SetBinChsClnFlag(fieldTp)
+	}
+	fieldTp.Decimal, fieldTp.Flen = decimal, flen
+	var tp evalTp
+	switch classType {
```

Comment:
There's mall differences between `fieldTp2EvalTp` and the `switch - case` here. `fieldTp.Decimal` needs to be set to 0 if `classType == ClassInt` and  `fieldTp.Charset, fieldTp.Collate ` need to be reset under certain condition